### ISSUE TITLE
[array-type] Incorrect fix when using an Array of arrow functions with rest parameters

### ISSUE THREAD

**Issue by cheit-epic** [typescript-eslint/typescript-eslint#4756]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/array-type": "error"
  }
}
```

```TS
const foo: Array<new (...args: any[]) => void> = [];
```
**Expected Result**

```TS
const foo: (new (...args: any[]) => void)[] = [];
```

**Actual Result**

```TS
const foo: new (...args: any[]) => void[] = [];
```

The lack of parentheses changes the type to be "a function that returns void[]" rather than "an array of functions that return void"

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.17.0` |
| `@typescript-eslint/parser`        | `5.17.0` |
| `TypeScript`                       | `4.4.4` |
| `ESLint`                           | `8.8.0` |
| `node`                             | `8.1.0` |

**Comment by bradzacher** [comment]
Same story with us not handling precedence correctly. We can reuse our new functions to help fix this.

**Comment by bradharms** [comment]
The title of this report may be wrong. The issue seems to be more with the type of an array of a constructor that takes a rest parameter:

`Array<new (...args: any[]) => any>` is corrected as `new (...args: any[]) => any[]` ❌ 

I don't see the problem occurring for an array of functions:

`Array<(...args: any[]) => any>` is corrected as `((...args: any[]) => any)[]` ✅

**Comment by armano2** [comment]
@bradharms correct, issue is present for type constructors `TSConstructorType`
